File: backend/mcp-code-intelligence-kotlin/bin/main/com/codeintel/memory/ingest/IngestPipeline.kt
/** Document ingestion pipeline — parse, chunk, embed, store. Enhanced with quality gate (KSA-110 F4). */
package com.codeintel.memory.ingest

import com.codeintel.log
import com.codeintel.memory.embedding.EmbeddingService
import com.codeintel.memory.ingest.autolink.AutoLinkResult
import com.codeintel.memory.ingest.autolink.AutoLinker
import com.codeintel.memory.models.KnowledgeEntry
import com.codeintel.memory.models.KnowledgeType
import com.codeintel.memory.models.MemoryTier
import com.codeintel.memory.repository.KnowledgeRepository
import com.codeintel.memory.search.IngestMeta
import com.codeintel.memory.search.QualityGate
import com.codeintel.memory.search.QualityResult

/** Result of ingesting a document. */
data class IngestResult(
    val entriesCreated: Int,
    val embeddingsGenerated: Int = 0,
    val source: String,
    val entryIds: List<Long> = emptyList(),
    val skipped: Boolean = false,
    val skipReason: String? = null,
    val autoLinkResult: AutoLinkResult? = null
)

/** Error thrown when quality gate rejects content. */
class QualityRejectionException(val quality: QualityResult) :
    RuntimeException(quality.message ?: "Quality gate rejected content")

class IngestPipeline(
    private val knowledgeRepo: KnowledgeRepository,
    private val embeddingService: EmbeddingService?,
    private val chunker: ChunkingStrategy = SemanticChunker()
) {
    private var qualityGate: QualityGate? = null
    private var autoLinker: AutoLinker? = null
    private var contradictionResolver: com.codeintel.memory.contradiction.ContradictionResolver? = null

    /** Inject QualityGate for ingest validation. */
    fun setQualityGate(gate: QualityGate) { this.qualityGate = gate }

    /** Inject AutoLinker for post-ingest linking (KSA-190). */
    fun setAutoLinker(linker: AutoLinker) { this.autoLinker = linker }

    /** Inject ContradictionResolver for detecting conflicting info on ingest. */
    fun setContradictionResolver(resolver: com.codeintel.memory.contradiction.ContradictionResolver) { this.contradictionResolver = resolver }

    /** Ingest a markdown document. */
    fun ingestMarkdown(text: String, source: String, type: String = KnowledgeType.CONTEXT.name): IngestResult {
        val doc = DocumentParser.parseMarkdown(text, source)
        return ingestParsedDocument(doc, type)
    }

    /** Ingest plain text. */
    fun ingestText(text: String, source: String, type: String = KnowledgeType.CONTEXT.name): IngestResult {
        val doc = DocumentParser.parsePlainText(text, source)
        return ingestParsedDocument(doc, type)
    }

    /** Ingest a single knowledge entry directly. Quality gate applied if set. */
    fun ingestEntry(content: String, summary: String, type: String, source: String? = null, tags: String = ""): Long {
        qualityGate?.let { gate ->
            val quality = gate.validate(content, IngestMeta(tags, type, source))
            if (quality.decision == "reject") throw QualityRejectionException(quality)
        }

        val entry = KnowledgeEntry(
            content = content,
            summary = summary,
            type = type,
            tier = MemoryTier.WORKING.name,
            source = source,
            tags = tags
        )
        val id = knowledgeRepo.insert(entry)
        embeddingService?.embedAndStore(id, summary)
        trySetQualityScore(id, content, tags, type, source)
        tryAutoLink(id)
        tryDetectContradiction(id)
        return id
    }

    /** Update agent_name for a knowledge entry after ingest. */
    fun updateAgentName(id: Long, agentName: String) {
        knowledgeRepo.updateAgentName(id, agentName)
    }

    /** Fire-and-forget auto-linking after ingest (KSA-190). */
    private fun tryAutoLink(entryId: Long): AutoLinkResult? {
        val linker = autoLinker ?: return null
        return try {
            linker.link(entryId)
        } catch (e: Exception) {
            log("[ingest] Auto-link failed for entry $entryId: $e")
            null
        }
    }

    /** Detect contradictions with existing entries after ingest. */
    private fun tryDetectContradiction(entryId: Long) {
        val resolver = contradictionResolver ?: return
        try {
            resolver.detectAndResolve(entryId)
        } catch (e: Exception) {
            log("[ingest] Contradiction detection failed for entry $entryId: $e")
        }
    }

    private fun trySetQualityScore(id: Long, content: String, tags: String, type: String, source: String?) {
        val gate = qualityGate ?: return
        try {
            val result = gate.validate(content, IngestMeta(tags, type, source))
            knowledgeRepo.updateQualityScore(id, result.score)
        } catch (_: Exception) { /* quality scoring must not break ingest */ }
    }

    private fun ingestParsedDocument(doc: ParsedDocument, type: String): IngestResult {
        var embeddingsGenerated = 0
        val source = doc.metadata["source"] ?: ""
        val allIds = mutableListOf<Long>()

        for (section in doc.sections) {
            if (section.content.isBlank()) continue
            val chunks = chunker.chunk(section.content, doc.metadata)
            for (chunk in chunks) {
                val id = storeChunk(chunk, section.heading, type, source)
                allIds.add(id)
                if (generateEmbedding(id, chunk.content)) embeddingsGenerated++
            }
        }
        // Auto-link last entry in batch (representative)
        val linkResult = if (allIds.isNotEmpty()) tryAutoLink(allIds.last()) else null
        log("Ingested $source: ${allIds.size} entries, $embeddingsGenerated embeddings")
        return IngestResult(allIds.size, embeddingsGenerated, source, allIds, autoLinkResult = linkResult)
    }

    private fun storeChunk(chunk: TextChunk, heading: String, type: String, source: String): Long {
        val summary = buildSummary(chunk.content, heading)
        val entry = KnowledgeEntry(
            content = chunk.content,
            summary = summary,
            type = type,
            tier = tierForType(type),
            source = source,
            sourceRef = "chunk:${chunk.index}",
            tags = heading
        )
        return knowledgeRepo.insert(entry)
    }

    private fun tierForType(type: String): String = when (type) {
        "REQUIREMENT", "ARCHITECTURE", "PROCEDURE", "API_DESIGN" -> MemoryTier.SEMANTIC.name
        "DECISION", "LESSON_LEARNED" -> MemoryTier.EPISODIC.name
        "ERROR_PATTERN" -> MemoryTier.PROCEDURAL.name
        else -> MemoryTier.WORKING.name
    }

    private fun buildSummary(content: String, heading: String): String {
        val firstLine = content.lines().firstOrNull { it.isNotBlank() } ?: ""
        val preview = firstLine.take(120)
        return if (heading.isNotBlank()) "$heading: $preview" else preview
    }

    private fun generateEmbedding(entryId: Long, text: String): Boolean {
        return embeddingService?.embedAndStore(entryId, text) ?: false
    }
}

File: backend/mcp-code-intelligence-nodejs/src/memory/ingest-pipeline.ts
/**
 * IngestPipeline — parse, chunk, and store knowledge entries.
 * Enhanced with quality gate validation before storage.
 * KSA-190: Added auto-linking after structured map extraction.
 */

import { KnowledgeRepository } from './knowledge-repo.js';
import { EmbeddingService } from './embedding/index.js';
import { parseMarkdown } from './document-parser.js';
import { SemanticChunker } from './chunking-strategy.js';
import { extractStructuredMap } from './structured-map-extractor.js';
import { EntityRepository } from './entity-repo.js';
import { classifyEntity } from './entity-classifier.js';
import type { QualityGate, QualityResult } from './v2/quality-gate.js';
import type { AutoLinker } from './auto-linker.js';
import type { AutoLinkResult } from './linking-strategies/types.js';
import type { ContradictionResolver, ResolutionResult } from './contradiction-resolver.js';

/** Result of ingesting a document. */
export interface IngestResult {
  entriesCreated: number;
  source: string;
}

/** Result of ingesting a single entry with quality info. */
export interface IngestEntryResult {
  id: number | null;
  quality: QualityResult | null;
  success: boolean;
  autoLink?: AutoLinkResult | null;
  contradiction?: ResolutionResult | null;
}

export class IngestPipeline {
  private readonly repo: KnowledgeRepository;
  private readonly embedding: EmbeddingService | null;
  private readonly chunker = new SemanticChunker(1024);
  private entityRepo: EntityRepository | null = null;
  private qualityGate: QualityGate | null = null;
  private autoLinker: AutoLinker | null = null;
  private contradictionResolver: ContradictionResolver | null = null;

  constructor(repo: KnowledgeRepository, embeddingService: EmbeddingService | null = null) {
    this.repo = repo;
    this.embedding = embeddingService;
  }

  /** Inject EntityRepository for structured map indexing. */
  setEntityRepo(repo: EntityRepository): void {
    this.entityRepo = repo;
  }

  /** Inject QualityGate for ingest validation. */
  setQualityGate(gate: QualityGate): void {
    this.qualityGate = gate;
  }

  /** Inject AutoLinker for automatic graph edge creation. KSA-190. */
  setAutoLinker(linker: AutoLinker): void {
    this.autoLinker = linker;
  }

  /** Inject ContradictionResolver for detecting conflicting info on ingest. */
  setContradictionResolver(resolver: ContradictionResolver): void {
    this.contradictionResolver = resolver;
  }

  /** Ingest a single knowledge entry with quality gate. Returns entry ID or rejection. */
  ingestEntry(content: string, summary: string, type: string, source?: string, tags = ''): number {
    // Quality gate check (if enabled)
    if (this.qualityGate) {
      const quality = this.qualityGate.validate(content, { tags, type, source });
      if (quality.decision === 'reject') {
        throw new QualityRejectionError(quality);
      }
    }

    const id = this.repo.insert({ content, summary, type, tier: 'WORKING', source, tags });
    this.tryEmbed(id, summary);
    this.tryExtractMap(id, content);
    this.tryAutoLink(id);
    this.trySetQualityScore(id, content, { tags, type, source });
    return id;
  }

  /** Ingest with full quality result returned. */
  ingestEntryWithQuality(
    content: string, summary: string, type: string, source?: string, tags = ''
  ): IngestEntryResult {
    if (this.qualityGate) {
      const quality = this.qualityGate.validate(content, { tags, type, source });
      if (quality.decision === 'reject') {
        return { id: null, quality, success: false };
      }
      const id = this.repo.insert({ content, summary, type, tier: 'WORKING', source, tags });
      this.tryEmbed(id, summary);
      this.tryExtractMap(id, content);
      const autoLink = this.tryAutoLink(id);
      this.trySetQualityScore(id, content, { tags, type, source });
      const contradiction = this.tryDetectContradiction(id);
      return { id, quality, success: true, autoLink, contradiction };
    }
    const id = this.repo.insert({ content, summary, type, tier: 'WORKING', source, tags });
    this.tryEmbed(id, summary);
    this.tryExtractMap(id, content);
    const autoLink = this.tryAutoLink(id);
    const contradiction = this.tryDetectContradiction(id);
    return { id, quality: null, success: true, autoLink, contradiction };
  }

  /** Ingest a markdown document — splits by sections. */
  ingestMarkdown(text: string, source: string, type = 'CONTEXT'): IngestResult {
    const doc = parseMarkdown(text, source);
    let entriesCreated = 0;
    for (const section of doc.sections) {
      if (!section.content.trim()) continue;
      const chunks = this.chunker.chunk(section.content);
      for (const chunk of chunks) {
        const summary = buildSummary(chunk.content, section.heading);
        const id = this.repo.insert({ content: chunk.content, summary, type, tier: tierForType(type), source, tags: section.heading });
        this.tryEmbed(id, summary);
        this.tryExtractMap(id, chunk.content);
        this.tryAutoLink(id);
        entriesCreated++;
      }
    }
    return { entriesCreated, source };
  }

  /** Ingest plain text. */
  ingestText(text: string, source: string, type = 'CONTEXT'): IngestResult {
    const chunks = this.chunker.chunk(text);
    let entriesCreated = 0;
    for (const chunk of chunks) {
      const summary = chunk.content.split('\n')[0]?.slice(0, 120) ?? source;
      const id = this.repo.insert({ content: chunk.content, summary, type, tier: tierForType(type), source, tags: '' });
      this.tryEmbed(id, summary);
      this.tryExtractMap(id, chunk.content);
      this.tryAutoLink(id);
      entriesCreated++;
    }
    return { entriesCreated, source };
  }

  /** Attempt to embed and store vector (fire-and-forget). */
  private tryEmbed(entryId: number, text: string): void {
    if (!this.embedding) return;
    this.embedding.embedAndStore(entryId, text).catch((err) => {
      process.stderr.write(`[ingest] Embed failed for entry ${entryId}: ${err}\n`);
    });
  }

  /** Extract structured map and index entities. */
  private tryExtractMap(entryId: number, content: string): void {
    try {
      const map = extractStructuredMap(content);
      const mapJson = JSON.stringify(map);
      this.repo.updateStructuredMap(entryId, mapJson);
      if (this.entityRepo && map.entities_mentioned.length > 0) {
        const entities = map.entities_mentioned.map(name => ({
          name,
          type: classifyEntity(name),
        }));
        this.entityRepo.indexEntities(entryId, entities);
      }
    } catch { /* extraction must not break ingest */ }
  }

  /** Auto-link entry to related entries (fire-and-forget). KSA-190. */
  private tryAutoLink(entryId: number): AutoLinkResult | null {
    if (!this.autoLinker) return null;
    try {
      return this.autoLinker.link(entryId);
    } catch (err) {
      process.stderr.write(`[ingest] Auto-link failed for entry ${entryId}: ${err}\n`);
      return null;
    }
  }

  /** Set quality score on entry after ingest. */
  private trySetQualityScore(
    entryId: number, content: string, meta: { tags?: string; type?: string; source?: string }
  ): void {
    if (!this.qualityGate) return;
    try {
      const result = this.qualityGate.validate(content, meta);
      this.repo.updateQualityScore(entryId, result.score);
    } catch { /* quality scoring must not break ingest */ }
  }

  /** Detect contradictions with existing entries (fire-and-forget). */
  private tryDetectContradiction(entryId: number): ResolutionResult | null {
    if (!this.contradictionResolver) return null;
    try {
      return this.contradictionResolver.detectAndResolve(entryId);
    } catch (err) {
      process.stderr.write(`[ingest] Contradiction detection failed for entry ${entryId}: ${err}\n`);
      return null;
    }
  }
}

/** Error thrown when quality gate rejects content. */
export class QualityRejectionError extends Error {
  readonly quality: QualityResult;
  constructor(quality: QualityResult) {
    super(quality.message ?? 'Quality gate rejected content');
    this.name = 'QualityRejectionError';
    this.quality = quality;
  }
}

/** Assign tier based on knowledge type. */
function tierForType(type: string): string {
  switch (type) {
    case 'REQUIREMENT': case 'ARCHITECTURE': case 'PROCEDURE': case 'API_DESIGN':
      return 'SEMANTIC';
    case 'DECISION': case 'LESSON_LEARNED': case 'ERROR_PATTERN':
      return 'EPISODIC';
    default:
      return 'WORKING';
  }
}

function buildSummary(content: string, heading: string): string {
  const firstLine = content.split('\n').find(l => l.trim()) ?? '';
  const preview = firstLine.slice(0, 120);
  return heading ? `${heading}: ${preview}` : preview;
}
